incubator/hnc/pkg/reconcilers/object.go

```diff
@@ -111,15 +120,22 @@ func (r *ObjectReconciler) SetMode(ctx context.Context, mode api.Synchronization
 	}
 	r.Log.Info("Changing mode of the object reconciler", "old", oldMode, "new", newMode)
 	r.Mode = newMode
-	// If the new mode is not "ignore", we need to update objects in the cluster
-	// (e.g., propagate or remove existing objects).
-	if newMode != api.Ignore {
-		err := r.enqueueAllObjects(ctx, r.Log)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
+	// For propagate and remove mode, we enqueue all objects to update objects in the cluster (e.g., propagate or
+	// remove existing objects). For ignore mode, we enqueue objects so that reconciliations of objects can trigger
+	// the config reconciler to update the status of the `config` singleton. Triggering object reconciliations in the
+	// ignore mode is necessary because only an object reconciler can compute NumPropagatedObjects for a given type.
```

Comment:
There might be some complications here. config reconciler can set the mode to anything (even an invalid mode). Object reconciler will validate the mode and change it to a valid mode if it is invalid (see [getValidateMode](https://github.com/kubernetes-sigs/multi-tenancy/blob/ca59e4a66a81d819183188e7d4c49f754302ff03/incubator/hnc/pkg/reconcilers/object.go#L129)). 

Ideally,  the config reconciler can get a valid mode from an object reconciler by adding& calling `GetMode` via the `TypeSyncer` interface. However, if a config reconciler creates (instead of syncs) an object reconciler, the `Mode` field of the object reconciler will be set to the valid mode when the (first) reconcile is [called](https://github.com/kubernetes-sigs/multi-tenancy/blob/ca59e4a66a81d819183188e7d4c49f754302ff03/incubator/hnc/pkg/reconcilers/object.go#L161). If config reconciler calls `GetMode` before any reconcile happens, it will still get an invalid mode.

This might not be an issue if we only want to know if the mode is `propagate` in config reconciler (check for `propagate` or unset mode), but it will be an issue if we want to know if the mode is `ignore`. To let config reconciler knows the valid mode, we can either move `getValidateMode` back to config reconciler so config reconciler will set a valid mode to an object reconciler. Or we can see if we can set the `Mode` field to a valid mode when creating the object reconciler, so that we don't need to wait for the first reconciler to happen.

WDYT?

Reply:
Discussed offline. I changed the method to a function.